replicator/src/main/scala/com/evolutiongaming/kafka/journal/replicator/TopicCommit.scala

```diff
@@ -1,44 +1,132 @@
+    RebalanceCallback.pure(())
+
+  /**
+   * Notify topic commit logic about partitions revoked during rebalance - called in
+   * [[com.evolutiongaming.skafka.consumer.RebalanceListener1.onPartitionsRevoked]].
+   * 
+   * Could be used to commit offsets for revoked partitions using
+   * [[RebalanceCallback.commit]] in order not to lose progress during rebalance.
+   *
+   * @param partitions revoked partitions
+   *
+   * @return [[RebalanceCallback]]
+   */
+  def onPartitionsRevoked(@nowarn("cat=unused") partitions: NonEmptySet[Partition]): RebalanceCallback[F, Unit] =
+    RebalanceCallback.pure(())
+
+  /**
+   * Notify topic commit logic about partitions lost - called in
+   * [[com.evolutiongaming.skafka.consumer.RebalanceListener1.onPartitionsLost]].
+   *
+   * When partitions are lost as opposed to being revoked:
+   * [[org.apache.kafka.clients.consumer.ConsumerRebalanceListener#onPartitionsLost(java.util.Collection)]]
+   * 
+   * Usually nothing could be done in this case apart from removing the lost partitions from internal state, if there
+   * is any.
+   *
+   * @param partitions lost partitions
+   *
+   * @return [[RebalanceCallback]]
+   */
+  def onPartitionsLost(@nowarn("cat=unused") partitions: NonEmptySet[Partition]): RebalanceCallback[F, Unit] =
+    RebalanceCallback.pure(())
 }
 
-object TopicCommit {
+private[journal] object TopicCommit {
 
-  def empty[F[_]: Applicative]: TopicCommit[F] = (_: Nem[Partition, Offset]) => ().pure[F]
+  def empty[F[_]: Applicative]: TopicCommit[F] = new EmptyTopicCommit[F]
 
+  @deprecated("renamed to sync", since = "4.2.0")
   def apply[F[_]](
     topic: Topic,
     metadata: String,
-    consumer: KafkaConsumer[F, _, _],
-  ): TopicCommit[F] = { (offsets: Nem[Partition, Offset]) =>
-    {
-      val offsets1 = offsets.mapKV { (partition, offset) =>
-        val offset1    = OffsetAndMetadata(offset, metadata)
-        val partition1 = TopicPartition(topic, partition)
-        (partition1, offset1)
-      }
-      consumer.commit(offsets1)
-    }
+    consumer: KafkaConsumer[F, ?, ?],
+  ): TopicCommit[F] = {
+    sync[F](
+      topic          = topic,
+      commitMetadata = metadata,
+      consumer       = consumer,
+    )
   }
 
+  /**
+   * [[TopicCommit]] which performs synchronous blocking commit on each offset mark call.
+   * 
+   * @param topic processed topic name
+   * @param commitMetadata metadata to pass to each consumer commit call
+   * @param consumer [[KafkaConsumer]] to call commit on
+   * @tparam F effect type
+   */
+  def sync[F[_]](
+    topic: Topic,
+    commitMetadata: String,
+    consumer: KafkaConsumer[F, ?, ?],
+  ): TopicCommit[F] = new SyncTopicCommit[F](
+    topic          = topic,
+    commitMetadata = commitMetadata,
+    consumer       = consumer,
+  )
+
+  @deprecated(
+    "use asyncPeriodic instead - the delayed impl blocks the pool-loop on commit and " +
```

Comment:
shouldn't `pool` be `poll`?
```suggestion
    "use `asyncPeriodic` instead - the `delayed` implementation blocks the poll-loop on commit and " +
```

Reply:
Will fix `pool`. Why do you think ticks are needed in deprecation messages? Is it how they do it in the Scala standard library? I don't see right away any value here.